### ISSUE TITLE
COMPILE ERROR: failed to create pipeline

### ISSUE THREAD

**Issue by xdobetter** [mitsuba-renderer/mitsuba3#612]
## Summary

Sorry to beather you, A great work! I want to opt geometry vertex use [Large Steps in Inverse Rendering of Geometry](https://github.com/rgl-epfl/large-steps-pytorch) in mitsuba3,but produce error. I don't know  whether large step code or mitsuba3 problem,so I want your help.

## System configuration

System information:

  OS: Ubuntu 20.04.5 LTS
  CPU: Intel Xeon Processor (Skylake, IBRS)
  GPU: Tesla V100S-PCIE-32GB
Tesla V100S-PCIE-32GB
  Python: 3.8.16 | packaged by conda-forge | (default, Feb  1 2023, 16:01:55) [GCC 11.3.0]
  NVidia driver: 520.61.05
  CUDA: 11.8.89
  LLVM: 12.0.0

  Dr.Jit: 0.4.1
  Mitsuba: 3.2.1
     Is custom build? False
     Compiled with: GNU 10.2.1
     Variants:
        scalar_rgb
        scalar_spectral
        cuda_ad_rgb
        llvm_ad_rgb

## Description

there will produce error when I add the code in opt part

```cpp
// add debug output image 
renderImg = mi.render(scene,params,spp=1,seed=it, seed_grad=it+1)
```

there is error details 

> COMPILE ERROR: failed to create pipeline                                                                                                                             
Info: Pipeline has 2 module(s), 19 entry function(s), 3 trace call(s), 0 continuation callable call(s), 8 direct callable call(s), 321 basic block(s) in entry functions, 10172 instruction(s) in entry functions, 14 non-entry function(s), 106 basic block(s) in non-entry functions, 1254 instruction(s) in non-entry functions, no debug information    
> Critical Dr.Jit compiler failure: jit_optix_check(): API error 7251 (OPTIX_ERROR_PIPELINE_LINK_ERROR): "Pipeline link error" in /project/ext/drjit-core/src/optix_cor
e.cpp:392.                                                                                                                                                           
Aborted (core dumped)  

The ablation experiment I've been trying so far:

- opt geometry (navie,aka no largeStep:) ) ----OK ; add debug output image ---- OK 
- opt geometry (largeStep) ----OK , but add debug output image will cause error. ---- Error

In opt geometry (largeStep) when don't add debug output image，there also have some informations , I guess that the source of this information is due to the implementation of large step algorithm.

nanobind: leaked 4 instances!
 nanobind: leaked 2 types!
> - leaked type "cholespy.CholeskySolverF"
> - leaked type "cholespy.MatrixType"

nanobind: leaked 2 functions!

>  - leaked function "__init__"
>  - leaked function "solve"

nanobind: this is likely caused by a reference counting issue in the binding code.

> drjit-autodiff: variable leak detected (3 variables remain in use)!
>  - variable a1504 (1 references)
>  - variable a2317 (1 references)
>  - variable a1648 (1 references)
> 

## Steps to reproduce

This is the code,the code can be run.

simple.xml from [mitsuba3 tutorial:Editing a scene](https://mitsuba.readthedocs.io/en/stable/src/rendering/editing_a_scene.html)

```python


import mitsuba as mi
import drjit as dr
import torch
import torch.nn as nn
from largesteps.optimize import AdamUniform
from largesteps.parameterize import from_differential,to_differential
from largesteps.geometry import compute_matrix
import numpy as np
from utils.file import *


mi.set_variant('cuda_ad_rgb')
scene=mi.load_file("./scenes/simple.xml")  

imageRef=mi.render(scene,spp=1024)
params=mi.traverse(scene)

keyPos = "teapot.vertex_positions"
keyFace = "teapot.faces"
keyNorm = "teapot.vertex_normals"

lambda_ = 19
v=params[keyPos]
f=params[keyFace]

vTen=torch.tensor(v,device='cuda')
vTen=torch.reshape(vTen,(vTen.shape[0]//3,3))

fTen=torch.tensor(list(f),device='cuda')
fTen=torch.reshape(fTen,(fTen.shape[0]//3,3))

M = compute_matrix(vTen, fTen, lambda_) 
u = to_differential(M, vTen) 

u.requires_grad=True
step_size = 3e-2

# opt=AdamUniform([u],step_size)
opt = torch.optim.Adam([u], lr=0.0002)


@dr.wrap_ad(source='torch',target='drjit')
def render(v,spp=256,seed=1):
    params[keyPos]=dr.ravel(v)
    params.update()
    return mi.render(scene,params,spp=spp,seed=seed,seed_grad=seed+1)

lossFn=nn.L1Loss()
iterationCount=2

for it in range(iterationCount):
    
    opt.zero_grad()

    vTransfer=from_differential(M,u)
   
    render_img=render(vTransfer,spp=16,seed=it)
    
    loss=lossFn(render_img,imageRef.torch())
    
    loss.backward() 
    
    opt.step()
    
    print(f"Iteration {it:02d}:  Loss = {loss.detach().cpu()} ",end="\n")
    # Debug output image
    renderImg = mi.render(scene,params,spp=1,seed=it, seed_grad=it+1) # add will cause error

print('\nOptimization complete.')
    
```

**Comment by njroussel** [comment]
Hi @xdobetter 

I can't tell you exactly why this is failing. Most likely some drjit/mitsuba bug. It's hard to trackdown because it will not happen with a debug build. Here are two solutions:

* You can move the `mi.render()`call before `loss.backward()`. This seems to work as intended.
* You can follow the excellent write-up over here: https://github.com/mitsuba-renderer/mitsuba3/discussions/600

**Comment by xdobetter** [comment]
NIce idea, Thank you very much. I'll try it right away

**Comment by njroussel** [comment]
Did it work? Can I close this issue?
